Problem 4.2

In [2]:
import  pandas as pd
import numpy as np

# creating matrix
M1 = np.random.randint (-10, 11, size=(3,3))
M2 = np.random.randint (-10, 11, size=(3,3))
M3 = np.random.randint (-10, 11, size=(3,3))

print(M1)
print(M2)
print(M3)

# inverses a matrix
M1_inverse = np.linalg.inv(M1)
M2_inverse = np.linalg.inv(M2)
M3_inverse = np.linalg.inv(M3)

print(M1_inverse)
print(M2_inverse)
print(M3_inverse)

# computes the product of a matrix
product1 = M1 @ M1_inverse
product2 = M2 @ M2_inverse
product3 = M3 @ M3_inverse

print(product1)
print(product2)
print(product3)



[[  1   6  -4]
 [ -1   0 -10]
 [ -3   0   3]]
[[ 2 -9 -6]
 [-7 -7 -7]
 [ 9 10 -5]]
[[-1  1  3]
 [-2  1  0]
 [-7 -3 -4]]
[[-0.         -0.09090909 -0.3030303 ]
 [ 0.16666667 -0.04545455  0.07070707]
 [-0.         -0.09090909  0.03030303]]
[[ 0.09259259 -0.09259259  0.01851852]
 [-0.08641975  0.03880071  0.04938272]
 [-0.00617284 -0.08906526 -0.06790123]]
[[-0.11428571 -0.14285714 -0.08571429]
 [-0.22857143  0.71428571 -0.17142857]
 [ 0.37142857 -0.28571429  0.02857143]]
[[ 1.00000000e+00  5.55111512e-17 -4.16333634e-17]
 [ 0.00000000e+00  1.00000000e+00  3.46944695e-17]
 [ 0.00000000e+00 -2.77555756e-17  1.00000000e+00]]
[[ 1.00000000e+00 -8.32667268e-17  0.00000000e+00]
 [-2.36789754e-16  1.00000000e+00 -8.32667268e-17]
 [ 5.29090660e-17 -1.38777878e-17  1.00000000e+00]]
[[ 1.00000000e+00 -2.77555756e-16  4.85722573e-17]
 [ 2.77555756e-17  1.00000000e+00  5.55111512e-17]
 [ 0.00000000e+00  2.22044605e-16  1.00000000e+00]]


Problem 5.1

In [4]:
df = pd.read_csv('kc_house_data.csv')

# drops columns id, date, and zipcode
df = df.drop(columns=['id', 'date', 'zipcode'])
df.head

summary_stats = pd.DataFrame({
    "average": df.mean(),
    "min": df.min(),
    "max": df.max(),
    "variance": df.var()
})

print(summary_stats)

lowest_avg = summary_stats["average"].idxmin()
highest_avg = summary_stats["average"].idxmax()

print(f"Lowest Average: {lowest_avg} = {summary_stats.loc[lowest_avg, 'average']:.2f}")
print(f"Highest Average: {highest_avg} = {summary_stats.loc[highest_avg, 'average']:.2f}")

lowest_var = summary_stats["variance"].idxmin()
highest_var = summary_stats["variance"].idxmax()

print(f"Lowest Variance: {lowest_var} = {summary_stats.loc[lowest_var, 'variance']:.2f}")
print(f"Highest Variance: {highest_var} = {summary_stats.loc[highest_var, 'variance']:.2f}")


                     average         min           max      variance
price          540088.141767  75000.0000  7.700000e+06  1.347824e+11
bedrooms            3.370842      0.0000  3.300000e+01  8.650150e-01
bathrooms           2.114757      0.0000  8.000000e+00  5.931513e-01
sqft_living      2079.899736    290.0000  1.354000e+04  8.435337e+05
sqft_lot        15106.967566    520.0000  1.651359e+06  1.715659e+09
floors              1.494309      1.0000  3.500000e+00  2.915880e-01
waterfront          0.007542      0.0000  1.000000e+00  7.485226e-03
view                0.234303      0.0000  4.000000e+00  5.872426e-01
condition           3.409430      1.0000  5.000000e+00  4.234665e-01
grade               7.656873      1.0000  1.300000e+01  1.381703e+00
sqft_above       1788.390691    290.0000  9.410000e+03  6.857347e+05
sqft_basement     291.509045      0.0000  4.820000e+03  1.958727e+05
yr_built         1971.005136   1900.0000  2.015000e+03  8.627973e+02
yr_renovated       84.402258      

Problem 5.2

In [6]:
correlations = df.corr()['price'].sort_values()
print(correlations)

positive_features = correlations[correlations > 0]
positive_features = positive_features.drop('price')

print("\nPositively Correlated Features:")
print(positive_features)



long             0.021626
condition        0.036362
yr_built         0.054012
sqft_lot15       0.082447
sqft_lot         0.089661
yr_renovated     0.126434
floors           0.256794
waterfront       0.266369
lat              0.307003
bedrooms         0.308350
sqft_basement    0.323816
view             0.397293
bathrooms        0.525138
sqft_living15    0.585379
sqft_above       0.605567
grade            0.667434
sqft_living      0.702035
price            1.000000
Name: price, dtype: float64

Positively Correlated Features:
long             0.021626
condition        0.036362
yr_built         0.054012
sqft_lot15       0.082447
sqft_lot         0.089661
yr_renovated     0.126434
floors           0.256794
waterfront       0.266369
lat              0.307003
bedrooms         0.308350
sqft_basement    0.323816
view             0.397293
bathrooms        0.525138
sqft_living15    0.585379
sqft_above       0.605567
grade            0.667434
sqft_living      0.702035
Name: price, dtype: float64


Which has the highest positive correlation with the response?

The feature with the highest positive correlation shown in the table is sqft_living with 0.702035.

Problem 5.3

No, I wasn't able to find any features with negative correlation coefficients.